sockeye/data_io.py

```diff
@@ -631,20 +625,25 @@ def prepare_data(source_fnames: List[str],
                  max_seq_len_target: int,
                  bucketing: bool,
                  bucket_width: int,
-                 samples_per_shard: int,
-                 min_num_shards: int,
+                 num_shards: int,
                  output_prefix: str,
                  bucket_scaling: bool = True,
                  keep_tmp_shard_files: bool = False,
-                 max_processes: int = 1):
+                 max_processes: int = 1,
+                 pool: multiprocessing.pool.Pool = None, 
+                 shards: List[Tuple[Tuple[str], Tuple[str]]] = None):
     logger.info("Preparing data.")
     # write vocabularies to data folder
     vocab.save_source_vocabs(source_vocabs, output_prefix)
     vocab.save_target_vocabs(target_vocabs, output_prefix)
 
     # Pass 1: get target/source length ratios.
-    length_statistics = analyze_sequence_lengths(source_fnames, target_fnames, source_vocabs, target_vocabs,
-                                                 max_seq_len_source, max_seq_len_target)
+    stats_args = ((source_path, target_path, source_vocabs, target_vocabs, max_seq_len_source, max_seq_len_target) for source_path, target_path in shards)
+    length_stats = pool.starmap(analyze_sequence_lengths, stats_args)
+    num_sents = sum([stat.num_sents for stat in length_stats])
+    length_ratio_mean = sum([stat.num_sents * stat.length_ratio_mean for stat in length_stats]) / num_sents
```

Comment:
The num_sents for each shard, since the sentences are currently being assigned to shards randomly, in an extreme case a shard could be empty. But the final num_sents would only be 0 if the input file is empty.